File: APIs/dataIngestAPI.py
# API to ingest data into a data transfer unit

__version__ = '0.0.1'

import os
import platform
import zmq
import logging
import cPickle
import traceback

class loggingFunction:
    def out (self, x, exc_info = None):
        if exc_info:
            print x, traceback.format_exc()
        else:
            print x
    def __init__ (self):
        self.debug    = lambda x, exc_info=None: self.out(x, exc_info)
        self.info     = lambda x, exc_info=None: self.out(x, exc_info)
        self.warning  = lambda x, exc_info=None: self.out(x, exc_info)
        self.error    = lambda x, exc_info=None: self.out(x, exc_info)
        self.critical = lambda x, exc_info=None: self.out(x, exc_info)


class noLoggingFunction:
    def out (self, x, exc_info = None):
        pass
    def __init__ (self):
        self.debug    = lambda x, exc_info=None: self.out(x, exc_info)
        self.info     = lambda x, exc_info=None: self.out(x, exc_info)
        self.warning  = lambda x, exc_info=None: self.out(x, exc_info)
        self.error    = lambda x, exc_info=None: self.out(x, exc_info)
        self.critical = lambda x, exc_info=None: self.out(x, exc_info)


def isWindows():
    returnValue = False
    windowsName = "Windows"
    platformName = platform.system()

    if platformName == windowsName:
        returnValue = True

    return returnValue


class dataIngest():
    # return error code
    def __init__ (self, useLog = False, context = None):

        if useLog:
            self.log = logging.getLogger("dataIngestAPI")
        elif useLog == None:
            self.log = noLoggingFunction()
        else:
            self.log = loggingFunction()

        # ZMQ applications always start by creating a context,
        # and then using that for creating sockets
        # (source: ZeroMQ, Messaging for Many Applications by Pieter Hintjens)
        if context:
            self.context    = context
            self.extContext = True
        else:
            self.context    = zmq.Context()
            self.extContext = False

        self.currentPID      = os.getpid()

        self.localhost       = "127.0.0.1"
        self.extIp           = "0.0.0.0"
        self.ipcPath         = "/tmp/zeromq-data-transfer"

        self.signalHost      = "zitpcx19282"
        self.signalPort      = "50050"

        # has to be the same port as configured in dataManager.conf as dataFetchPort
        self.eventDetPort    = "50003"
        #TODO add port in config
        # has to be the same port as configured in dataManager.conf as ...
        self.dataFetchPort   = "50010"

        self.signalConId     = "tcp://{ip}:{port}".format(ip=self.signalHost, port=self.signalPort)

        if isWindows():
            self.log.info("Using tcp for internal communication.")
            self.eventDetConId  = "tcp://{ip}:{port}".format(ip=self.localhost, port=self.dataFetchPort)
            self.dataFetchConId = "tcp://{ip}:{port}".format(ip=self.localhost, port=self.dataFetchPort)
        else:
            self.log.info("Using ipc for internal communication.")
            self.eventDetConId  = "ipc://{path}/{id}".format(path=self.ipcPath, id="eventDetConId")
            self.dataFetchConId = "ipc://{path}/{id}".format(path=self.ipcPath, id="dataFetchConId")
#            self.eventDetConId   = "ipc://{path}/{pid}_{id}".format(path=self.ipcPath, pid=self.currentPID, id="eventDetConId")
#            self.dataFetchConId  = "ipc://{path}/{pid}_{id}".format(path=self.ipcPath, pid=self.currentPID, id="dataFetchConId")

        self.signalSocket    = None
        self.eventDetSocket  = None
        self.dataFetchSocket = None

        self.poller              = zmq.Poller()

        self.filename            = False
        self.openFile            = False
        self.filePart            = None

        self.responseTimeout     = 1000

        self.__createSocket()


    def __createSocket (self):

        # To send file open and file close notification, a communication socket is needed
        self.signalSocket = self.context.socket(zmq.REQ)

        # time to wait for the sender to give a confirmation of the signal
#        self.signalSocket.RCVTIMEO = self.responseTimeout
        try:
            self.signalSocket.connect(self.signalConId)
            self.log.info("signalSocket started (connect) for '" + self.signalConId + "'")
        except Exception as e:
            self.log.error("Failed to start signalSocket (connect): '" + self.signalConId + "'", exc_info=True)
            raise

        # using a Poller to implement the signalSocket timeout (in older ZMQ version there is no option RCVTIMEO)
#        self.poller = zmq.Poller()
        self.poller.register(self.signalSocket, zmq.POLLIN)


        self.eventDetSocket = self.context.socket(zmq.PUSH)
        try:
            self.eventDetSocket.connect(self.eventDetConId)
            self.log.info("eventDetSocket started (connect) for '" + self.eventDetConId + "'")
        except:
            self.log.error("Failed to start eventDetSocket (connect): '" + self.eventDetConId + "'", exc_info=True)
            raise

        self.dataFetchSocket  = self.context.socket(zmq.PUSH)
        try:
            self.dataFetchSocket.connect(self.dataFetchConId)
            self.log.info("dataFetchSocket started (connect) for '" + self.dataFetchConId + "'")
        except:
            self.log.error("Failed to start dataFetchSocket (connect): '" + self.dataFetchConId + "'", exc_info=True)
            raise


    # return error code
    def createFile (self, filename):
        if self.openFile and self.openFile != filename:
            raise Exception("File " + str(filename) + " already opened.")

        # send notification to receiver
        self.signalSocket.send("OPEN_FILE")
        self.log.info("Sending signal to open a new file.")

        message = self.signalSocket.recv()
        self.log.debug("Received responce: " + str(message))

        self.filename = filename
        self.filePart = 0


    def write (self, data):
        # send event to eventDet
#        message = {
#                "filename" : self.filename,
#                "filePart" : self.filePart
#                }
#        message = "{ 'filename': " + self.filename + ", 'filePart': " + self.filePart + "}"
        message = '{ "filePart": ' + str(self.filePart) + ', "filename": "' + self.filename + '" }'
        self.eventDetSocket.send(message)

        # send data to ZMQ-Queue
        self.dataFetchSocket.send(data)
        self.filePart += 1


    # return error code
    def closeFile (self):
        # send close-signal to signal socket
        sendMessage = "CLOSE_FILE"
        try:
            self.signalSocket.send(sendMessage)
            self.log.info("Sending signal to close the file to signalSocket.")
        except:
            raise Exception("Sending signal to close the file to signalSocket...failed.")

        # send close-signal to event Detector
        try:
            self.eventDetSocket.send(sendMessage)
            self.log.debug("Sending signal to close the file to eventDetSocket. (sendMessage=" + sendMessage + ")")
        except:
            raise Exception("Sending signal to close the file to eventDetSocket...failed.")

        try:
            socks = dict(self.poller.poll(10000)) # in ms
        except:
            socks = None
            self.log.error("Could not poll for signal", exc_info=True)

        # if there was a response
        if socks and self.signalSocket in socks and socks[self.signalSocket] == zmq.POLLIN:
            self.log.info("Received answer to signal...")
            #  Get the reply.
            recvMessage = self.signalSocket.recv()
            self.log.info("Received answer to signal: " + str(recvMessage) )
        else:
            recvMessage = None

        if recvMessage != sendMessage:
            self.log.debug("recieved message: " + str(recvMessage))
            self.log.debug("send message: " + str(sendMessage))
            raise Exception("Something went wrong while notifying to close the file")

        self.openFile = None
        self.filePart = None


    ##
    #
    # Send signal that the displayer is quitting, close ZMQ connections, destoying context
    #
    ##
    def stop (self):
        try:
            if self.signalSocket:
                self.log.info("closing signalSocket...")
                self.signalSocket.close(linger=0)
                self.signalSocket = None
            if self.eventDetSocket:
                self.log.info("closing eventDetSocket...")
                self.eventDetSocket.close(linger=0)
                self.eventDetSocket = None
            if self.dataFetchSocket:
                self.log.info("closing dataFetchSocket...")
                self.dataFetchSocket.close(linger=0)
                self.dataFetchSocket = None
        except:
            self.log.error("closing ZMQ Sockets...failed.", exc_info=True)

        # if the context was created inside this class,
        # it has to be destroyed also within the class
        if not self.extContext and self.context:
            try:
                self.log.info("Closing ZMQ context...")
                self.context.destroy(0)
                self.context = None
                self.log.info("Closing ZMQ context...done.")
            except:
                self.log.error("Closing ZMQ context...failed.", exc_info=True)


    def __exit__ (self):
        self.stop()


    def __del__ (self):
        self.stop()

File: APIs/dataIngestAPI.c
// API to ingest data into a data transfer unit

#include <dataIngestAPI.h>
#include <zmq.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>
#include <errno.h>
#include <json.h>


// helper functions for sending and receiving messages
// source: Pieter Hintjens: ZeroMQ; O'Reilly
static char* s_recv (void *socket)
{
    zmq_msg_t message;
    zmq_msg_init (&message);
    int size = zmq_msg_recv (&message, socket, 0);
    if (size == -1)
        return NULL;
    char *string = malloc (size + 1);
    memcpy (string, zmq_msg_data (&message), size);
//    printf("recieved string: %s\n", string);
    zmq_msg_close (&message);
    string [size] = 0;
    return string;
};


static int s_send (void * socket, const char* string, int len)
{
    int bytesSent;
    int rc;

    /* Create a new message, allocating bytes for message content */
    zmq_msg_t message;
    printf("Sent message of size: %i\n", len);
    rc = zmq_msg_init_size (&message, len);
    assert (rc == 0);

    /* Fill in message content */
    memcpy (zmq_msg_data (&message), string, len);

    /* Send the message to the socket */
    bytesSent = zmq_msg_send (&message, socket, 0);
    assert (bytesSent == len);

    return bytesSent;
}


/*
static int s_send (void * socket, char* string)
{
//    printf("Sending: %s\n", string);
    int size = zmq_send (socket, string, strlen (string), 0);
    if (size == -1)
    {
        fprintf(stderr, "Failed to send message (message: '%s'): %s\n",
                string, strerror( errno ));
//        fprintf (stderr, "ERROR: Sending failed\n");
//        perror("");
    }

    return size;
}
*/

struct dataIngest {

    char *localhost;
    char *extIp;
    char *ipcPath;

    char *signalHost;
    char *signalPort;

    char *eventDetPort;
    char *dataFetchPort;

    //  Prepare our context and socket
    void *context;
    void *fileOpSocket;
    void *eventDetSocket;
    void *dataFetchSocket;

    //  Initialize poll set
//    zmq::pollitem_t items [];

    char *filename;
    int  openFile;
    int  filePart;

    int  responseTimeout;

};


HIDRA_ERROR dataIngest_init (dataIngest **out)
{
    dataIngest* dI = malloc(sizeof(dataIngest));

    *out = NULL;

    dI->localhost       = "localhost";
    dI->extIp           = "0.0.0.0";
    dI->ipcPath         = "/tmp/HiDRA";

    dI->signalHost      = "zitpcx19282";
    dI->signalPort      = "50050";

    // has to be the same port as configured in dataManager.conf as eventDetPort
    dI->eventDetPort    = "50003";
    // has to be the same port as configured in dataManager.conf as dataFetchPort
    dI->dataFetchPort   = "50010";

    char signalConId[128];
    char eventDetConId[128];
    char dataFetchConId[128];

    snprintf(signalConId, sizeof(signalConId), "tcp://%s:%s", dI->signalHost, dI->signalPort);

    // if plattform is Windows:
//    snprintf(eventDetConId, sizeof(eventDetConId), "tcp://%s:%s", dI->localhost, dI->eventDetPort);
//    snprintf(dataFetchConId, sizeof(dataFetchConId), "tcp://%s:%s", dI->localhost, dI->dataFetchPort);
    // else:
    snprintf(eventDetConId, sizeof(eventDetConId), "ipc://%s/eventDet", dI->ipcPath);
    snprintf(dataFetchConId, sizeof(dataFetchConId), "ipc://%s/dataFetch", dI->ipcPath);

    // Initialize sockets
    if ( (dI->context = zmq_ctx_new ()) == NULL )
    {
		perror("Cannot create 0MQ context: ");
        return ZMQERROR;
	}

	if ( (dI->fileOpSocket = zmq_socket (dI->context, ZMQ_REQ)) == NULL )
    {
		perror("Could not create 0MQ fileOpSocket");
        return ZMQERROR;
	}

	if ( (dI->eventDetSocket = zmq_socket (dI->context, ZMQ_PUSH)) == NULL )
    {
		perror("Could not create 0MQ eventDetSocket");
        return ZMQERROR;
	}

    if ( (dI->dataFetchSocket = zmq_socket (dI->context, ZMQ_PUSH)) == NULL )
    {
		perror("Could not create 0MQ dataFetchSocket: ");
        return ZMQERROR;
	}

    dI->filePart        = 0;
    dI->responseTimeout = 1000;

    char connectionStr[128];
    int rc;

    // Create sockets
    if ( zmq_connect(dI->fileOpSocket, signalConId) )
    {
        fprintf(stderr, "Failed to start fileOpSocket (connect) for '%s': %s\n",
                signalConId, strerror( errno ));
        //TODO stop socket
        return ZMQERROR;
    }
    else
    {
        printf("fileOpSocket started (connect) for '%s'\n", signalConId);
    }

    if ( zmq_connect(dI->eventDetSocket, eventDetConId) )
    {
        fprintf (stderr, "Failed to start eventDetSocket (connect) for '%s': %s\n",
                eventDetConId, strerror( errno ));
        //TODO stop socket
        return ZMQERROR;
    }
    else
    {
        printf("eventDetSocket started (connect) for '%s'\n", eventDetConId);
    }

    if ( zmq_connect(dI->dataFetchSocket, dataFetchConId) )
    {
        fprintf (stderr, "Failed to start dataFetchSocket (connect) for '%s': %s\n",
                dataFetchConId, strerror( errno ));
        //TODO stop socket
        return ZMQERROR;
    }
    else
    {
        printf("dataFetchSocket started (connect) for '%s'\n", dataFetchConId);
    }

    *out = dI;

    return SUCCESS;

}


HIDRA_ERROR dataIngest_createFile (dataIngest *dI, char *fileName)
{

//        if self.openFile and self.openFile != filename:
//            raise Exception("File " + str(filename) + " already opened.")

    char *message;
    int rc;

    // Send notification to receiver
    message = "OPEN_FILE";
    rc = s_send (dI->fileOpSocket, message, strlen(message));
    if (rc == -1) return COMMUNICATIONFAILED;
    printf ("Sending signal to open a new file.\n");

    message = s_recv (dI->fileOpSocket);
    printf ("Received responce: '%s'\n", message);

    dI->filename = fileName;
    dI->filePart = 0;

    free (message);

    return SUCCESS;
}


HIDRA_ERROR dataIngest_write (dataIngest *dI, char *data, int size)
{

    char message[128];
    int rc;
    json_object * metadata_json;
    const char *metadata_string;


    metadata_json = json_object_new_object();

    json_object_object_add(metadata_json,"filename", json_object_new_string(dI->filename));
    json_object_object_add(metadata_json,"filePart", json_object_new_int(dI->filePart));
    json_object_object_add(metadata_json,"chunkSize", json_object_new_int(size));

    metadata_string = json_object_to_json_string ( metadata_json );

    // Send event to eventDetector
    rc = s_send (dI->eventDetSocket, metadata_string, strlen(metadata_string));
    if (rc == -1) return COMMUNICATIONFAILED;

    json_object_put ( metadata_json );

    // Send data to ZMQ-Queue
    rc = s_send (dI->dataFetchSocket, data, size);
    if (rc == -1) return COMMUNICATIONFAILED;

    printf ("Writing: %s\n", message);

    dI->filePart += 1;

    return SUCCESS;
};


HIDRA_ERROR dataIngest_closeFile (dataIngest *dI)
{

    char *message = "CLOSE_FILE";
    char *answer;

    int rc;

    // Send close-signal to signal socket
    rc = s_send (dI->fileOpSocket, message, strlen(message));
    if (rc == -1) return COMMUNICATIONFAILED;
    //        perror("Sending signal to close the file to fileOpSocket...failed.")
    printf ("Sending signal to close the file to fileOpSocket  (sendMessage=%s)\n", message);


    // send close-signal to event Detector
    rc = s_send (dI->eventDetSocket, message, strlen(message));
    if (rc == -1) return COMMUNICATIONFAILED;
//        perror("Sending signal to close the file to eventDetSocket...failed.)")
    printf ("Sending signal to close the file to eventDetSocket (sendMessage=%s)\n", message);


//    try:
//        socks = dict(self.poller.poll(10000)) # in ms
//    except:
//        socks = None
//        self.log.error("Could not poll for signal", exc_info=True)
//
//    // if there was a response
//    if socks and self.fileOpSocket in socks and socks[self.fileOpSocket] == zmq.POLLIN:
    // Get the reply.
    answer = s_recv (dI->fileOpSocket);
    printf ("Received answer to signal: %s\n", answer);

    if ( strcmp(message,answer) != 0 )
    {
        perror ("Something went wrong while notifying to close the file");
        printf ("recieved message: %s\n", answer);
        printf ("send message: %s\n", message);
    };

    dI->filename = "";
    dI->filePart = 0;

    free (answer);

    return SUCCESS;
};

HIDRA_ERROR dataIngest_stop (dataIngest *dI)
{

    printf ("closing fileOpSocket...\n");
    zmq_close(dI->fileOpSocket);
    printf ("closing eventDetSocket...\n");
    zmq_close(dI->eventDetSocket);
    printf ("closing dataFetchSocket...\n");
    zmq_close(dI->dataFetchSocket);
//    perror("closing ZMQ Sockets...failed.")

    printf ("Closing ZMQ context...\n");
    zmq_ctx_destroy(dI->context);
//    perror("Closing ZMQ context...failed.")

    free (dI);
    printf ("Cleanup finished.\n");

    return SUCCESS;
};
